numba/np/arrayobj.py

```diff
@@ -3829,6 +3829,10 @@ def make_array_nditer(context, builder, sig, args):
     nditerty = sig.return_type
     arrtys = nditerty.arrays
 
+    if any(_arrty.layout not in 'CF' for _arrty in arrtys):
+        raise TypeError('Only C or F contiguous arrays are '
+                        'accepted by Numba implementation of np.nditer')
```

Comment:
This check is being made at compile time? What if the array is typed as `A` ordered but has a runtime layout of `F`? Example:
```python
from numba import njit, typeof
import numpy as np

data = np.array([[1., 3, 2, 7,],
                 [-5, 4, 2, 3,],
                 [9, -3, 1, 1,],
                 [2, -2, 2, 8,]], order='F')

reshaped_data = data.reshape((1, 4, 4))

@njit
def foo(arr):
    tmp = arr[0]
    acc = 0.
    for ix in np.nditer(tmp):
        acc += np.take(ix, 0)
    return acc

foo(reshaped_data)

foo.inspect_types()
```
Also, if an assertion is being made about supported types it needs to be done during type inference, not lowering.

Reply:
IIRC `np.ndarray.flags.{c, f}_contiguous` works in `nopython` mode, had to do something similar here #8349.